packages/amplify_core/lib/src/config/auth/cognito/user_pool.dart

```diff
@@ -15,13 +15,16 @@ class CognitoUserPoolConfig
   final CognitoOAuthConfig? hostedUI;
   final String? endpoint;
 
-  const CognitoUserPoolConfig({
+  String? pinpointEndpointId;
+
+  CognitoUserPoolConfig({
```

Comment:
These classes are a direct representation of the `amplifyconfiguration.dart` file, so this is not the right place for this.

Data classes should also be immutable. We override `==` and `hashCode` on all data classes with `AWSEquatable`, which is marked with `@immutable` to ensure these values never change over the lifetime of the object.

This is why the analyzer triggers an error for this line: 

```
 info - lib/src/config/auth/cognito/user_pool.dart:9:7 - This class (or a class that this class inherits from) is marked as '@immutable', but one or more of its instance fields aren't final: CognitoUserPoolConfig.pinpointEndpointId - must_be_immutable
```

The reasons for this are two-fold:

- Mutating data classes over time leads to difficult-to-reproduce errors and hard-to-follow data flow. This is why functional programming languages expressly prohibit this paradigm.
- Anytime `==` or `hashCode` are overridden for structural equality, they _must_ never be mutated. For example, imagine putting a CognitoUserPoolConfig in a HashSet, then setting the pinpointEndpointId value - the HashSet would tell you it's not there! This is pretty confusing without having created a new object.

Reply:
Thanks for the great insights here on that.  I've changed the code to add a analyticsMetadataProvider in the stateMachineManager's dependencyManager.